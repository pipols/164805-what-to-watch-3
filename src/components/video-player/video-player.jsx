import React from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import withVideoPlayer from "../../hocs/with-video-player/with-video-player.jsx";
import {getFilm} from "../../reducer/state/selector";
import Preloader from "../preloader/preloader.jsx";
import history from '../../history';

const VideoPlayer = (props) => {
  const {isPlay, onTimeUpdate, setDuration, progress, duration, onPlayClick, onFullscreenClick, forwardedRef, film} = props;

  return props.film
    ? <div className="player">
      <video
        ref={forwardedRef}
        onTimeUpdate={onTimeUpdate}
        onLoadedMetadata={setDuration}
        src={film.videoLink}
        className="player__video"
        poster={film.previewImage}>
      </video>

      <button onClick={() => history.push(`/film/${film.id}`)} type="button" className="player__exit">Exit</button>

      <div className="player__controls">
        <div className="player__controls-row">
          <div className="player__time">
            <progress className="player__progress" value={progress} max="100"></progress>
            <div className="player__toggler" style={{left: `${progress}%`}}>Toggler</div>
          </div>
          <div className="player__time-value">{duration}</div>
        </div>

        <div className="player__controls-row">
          <button onClick={onPlayClick} type="button" className="player__play">
            {isPlay || <><svg viewBox="0 0 19 19" width="19" height="19">
              <use xlinkHref="#play-s"></use>
            </svg>
              <span>Play</span></>}

            {isPlay && <><svg viewBox="0 0 14 21" width="14" height="21">
              <use xlinkHref="#pause"></use>
            </svg>
              <span>Pause</span></>}
          </button>
          <div className="player__name">{film.title}</div>
          <button onClick={onFullscreenClick} type="button" className="player__full-screen">
            <svg viewBox="0 0 27 27" width="27" height="27">
              <use xlinkHref="#full-screen"></use>
            </svg>
            <span>Full screen</span>
          </button>
        </div>
      </div>
    </div>

    : <Preloader />;
};

VideoPlayer.propTypes = {
  isPlay: PropTypes.bool.isRequired,
  onTimeUpdate: PropTypes.func.isRequired,
  setDuration: PropTypes.func.isRequired,
  onPlayClick: PropTypes.func.isRequired,
  onFullscreenClick: PropTypes.func.isRequired,
  forwardedRef: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({current: PropTypes.any})
  ]),
  progress: PropTypes.number.isRequired,
  duration: PropTypes.string.isRequired,
  film: PropTypes.shape({
    title: PropTypes.string.isRequired,
    cover: PropTypes.string.isRequired,
    previewImage: PropTypes.string.isRequired,
    poster: PropTypes.string.isRequired,
    backgroundColor: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    rating: PropTypes.number.isRequired,
    votes: PropTypes.number.isRequired,
    producer: PropTypes.string.isRequired,
    actors: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
    duration: PropTypes.number.isRequired,
    genre: PropTypes.string.isRequired,
    year: PropTypes.number.isRequired,
    id: PropTypes.number.isRequired,
    isFavorite: PropTypes.bool.isRequired,
    videoLink: PropTypes.string.isRequired,
    preview: PropTypes.string.isRequired,
  }),
};

const mapStateToProps = (state) => ({
  film: getFilm(state)
});

const VideoPlayerWrap = React.forwardRef((props, ref) => <VideoPlayer {...props} forwardedRef={ref} />);

export {VideoPlayer};
export default connect(mapStateToProps)(withVideoPlayer(React.memo(VideoPlayerWrap)));
