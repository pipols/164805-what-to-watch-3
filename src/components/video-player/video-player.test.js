import React from "react";
import renderer from "react-test-renderer";
import {VideoPlayer} from "./video-player.jsx";

const film = {
  title: `The Grand Budapest Hotel`,
  cover: `the-grand-budapest-hotel-poster.jpg`,
  previewImage: `the-grand-budapest-hotel-poster.jpg`,
  poster: `bg-the-grand-budapest-hotel.jpg`,
  backgroundColor: `#fff`,
  description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H.`,
  rating: 8.9,
  votes: 240,
  producer: `Wes Andreson`,
  actors: [
    `Bill Murray`,
    `Edward Norton`,
    `Jude Law`,
    `Willem Dafoe`,
    `Saoirse Ronan`,
    `Tony Revoloru`,
    `Tilda Swinton`,
    `Tom Wilkinson`,
    `Owen Wilkinson`,
    `Adrien Brody`,
    `Ralph Fiennes`,
    `Jeff Goldblum`
  ],
  duration: 120,
  genre: `Drama`,
  year: 2014,
  id: 1,
  isFavorite: true,
  videoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  preview: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
};

it(`<VideoPlayer /> renders correctly`, () => {
  const tree = renderer
  .create(<VideoPlayer
    isPlay={true}
    onTimeUpdate={() => {}}
    setDuration={() => {}}
    progress={100}
    duration={100}
    onPlayClick={() => {}}
    onFullscreenClick={() => {}}
    forwardedRef={() => {}}
    film={film}
  />)
  .toJSON();

  expect(tree).toMatchSnapshot();
});
