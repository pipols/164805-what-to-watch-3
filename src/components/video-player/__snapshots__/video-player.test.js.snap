// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VideoPlayer /> renders correctly 1`] = `
<video
  autoPlay={true}
  height="175"
  muted={true}
  poster="fantastic-beasts-the-crimes-of-grindelwald.jpg"
  src="test.ru"
  width="280"
/>
`;
